bsp4j/src/main/java/ch/epfl/scala/bsp4j/JvmExtension.xtend

```diff
@@ -30,6 +24,30 @@ class JvmEnvironmentItem {
   }
 }
 
+@JsonRpcData
+class JvmEnvironmentParams {
+  @NonNull List<BuildTargetIdentifier> targets
+  new(@NonNull List<BuildTargetIdentifier> targets) {
+    this.targets = targets
+  }
+}
+
+@JsonRpcData
+class JvmEnvironmentResult {
+  @NonNull List<JvmEnvironmentItem> items
+  new(@NonNull List<JvmEnvironmentItem> items) {
+    this.items = items
```

Comment:
I don't think the `jvmEnvironment` should include a classpath and I think it can be per-workaspace.

```scala
case class JvmEnvironmentParams()
case class JvmEnvironmentResult(javaHome: String, javaVersion: String)
```

What do you think @jastice ?

Reply:
That was my initial implementation, yet Justin mentioned that in general jdk can be different per target. We will be able to use JvmEnvironment for running apps like we use JvmTestEnvironment for running tests.

Reply:
That's fair. I don't think the response should include the classpath however, this method should be cheap to respond to. I wonder if it's also worth adding the ability to say the JDK is the same for all targets 🤔

Reply:
Hmm, maybe we should put jdk and other runtime parameters in separate endpoints. Not sure how to name the new one. It could respond with say:
```
case class JdkResponse(projectJdk: Option[JdkItem], moduleJdks: List[JdkItem])
case class JdkItem(buildTargetId: Option[BuildTargetId], javaHome: String, javaVersion: String)
```

Reply:
Having the classpath in there has the benefit that we effectively get Java support for non-Scala modules

What's the reason this method in particular should be cheap? I'm not against splitting it to separate endpoints, just wondering in what cases it would be beneficial.

Reply:
I believe I'm misunderstanding the information needed in IntelliJ to configure the JDK. My understanding is that this issue is to avoid the manual prompt to configure the "Project JDK"

![Screenshot 2020-03-19 at 17 41 07](https://user-images.githubusercontent.com/1408093/77674720-e9ac3f00-6f83-11ea-840b-dfaed9d54189.png)

Is the goal is to support non-Scala targets instead?

Reply:
The prompt is due to a missing module JDK configuration and a bug in the code that generates the prompt. I'm fixing this without any changes to BSP.

Adding this change to BSP however allows us to configure the JDK explicitly from the server rather than best-guessing or manually configuring it on the client side, which can be necessary if there is a version incompatibility.

Supporting Java-only targets is an added benefit, but a rather important one.

Reply:
>  I'm fixing this without any changes to BSP.

How does IntelliJ infer the default Java home? The Java version influences the presentation compiler to resolve symbols like `java.nio.file.Files`. We should make sure that IntelliJ doesn't for example expose JDK 11 symbols when the build is on JDK 8.

The `jvmTestEnvironment` endpoint is only called right before running tests so I'm not sure if it's a good candidate to configure the presentation compiler.

Would it make sense to include the Java home and java version in a `JavaBuildTarget` the same way we do for `ScalaBuildTarget`? https://build-server-protocol.github.io/docs/extensions/scala.html We can make `ScalaBuildTarget` extend `JavaBuildTarget`. That way we can add an explicit Java extension to the protocol.

Reply:
By default I just take the most recent one available. That of course has the problem you mention: it might expose methods you don't want, or be downright incompatible, This endpoint addresses that case.

`JavaBuildTarget` would be an alternative. On the other hand, these parameters would also apply to other JVM languages. I don't have a strong opinion on this yet.

Calling it before tests doesn't mean we can't call it any other time we need it.

Reply:
> Calling it before tests doesn't mean we can't call it any other time we need it.

How would that work when the source code doesn't compile successfully? `jvmTestEnvironment` only returns if the code compiles without errors.

Reply:
The reason I think it should be cheap to query the java home/version is that 1) this version is cheap to produce and 2) we need this information for the presentation compiler before triggering "build project"